A python script to wrap image read and write for Nifti Images.

Unique-Usman opened the pull request:
### Describe the changes you have made in this PR
I am using this script as the placeholder for the final script, we can continue the discussion from here. 
@etpeterson 

I have quite not understand the available algorithm and where they are located. Kindly give more information. Thank you.

Fixes #58

### Checklist

- [ ] Self-review of changed code
- [ ] Added automated tests where applicable
- [ ] Update Docs & Guides

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,64 @@
+import nibabel as nib
+import numpy as np
+
+def read_nifti_4d(input_file):
```
This is called 4d, but could be any dimensionality.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,64 @@
+    For reading the 4d nifti image
+    """
+    nifti_img = nib.load(input_file)
+    return nifti_img.get_fdata()
```
We'll also want the header information. Maybe return both?

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,64 @@
+    """
+    For saving the 3d nifti images of the output of the algorithm
+    """
+    output_img = nib.Nifti1Image(data, np.eye(4))
```
Creating this output image might necessitate copying aspects of the input image. This function should probably accept or at least pass through arguments to `Nifti1Image`. https://nipy.org/nibabel/reference/nibabel.nifti1.html#nibabel.nifti1.Nifti1Image

Reasonable defaults for whatever n-dimensional image make sense though.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,64 @@
+        bvecs = read_json_file(bvec_file)
+        bvals = read_json_file(bval_file)
+
+        # Pass additional arguments to the algorithm
```
Initialize the fitting like [this](https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/blob/main/tests/IVIMmodels/unit_tests/test_ivim_synthetic.py#L27).

And the fitting is [here](https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/blob/main/tests/IVIMmodels/unit_tests/test_ivim_synthetic.py#L40).

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,64 @@
+
+        # Pass additional arguments to the algorithm
+        # Passing the values to the selectect algorithm and saving it
+
```
And don't forget the saving of the 3 images.

etpeterson commented:
> I have quite not understand the available algorithm and where they are located. Kindly give more information. Thank you.

Does it make more sense with my comments?

etpeterson commented:
@IvanARashid This could use your multi-voxel fitting. Also, if you have ideas about how to best merge with the wrapper and handle inputs/outputs of the wrapper, that would be helpful.

Unique-Usman commented:
Kindly help on the https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/pull/60/commits/6c779170ea70b9d4b98617d04d9043b97c87e995#diff-aea94673dea889894c34fe41c94d6cb8fb0d3373e0677947893ba1e79f4359edR71

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -2,9 +2,12 @@
+import random
 
-def read_nifti_4d(input_file):
+def read_nifti_file(input_file):
```
This should also return the header or just the complete nifti for the user to separate.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -27,11 +30,11 @@ def read_json_file(json_file):
     For saving the 3d nifti images of the output of the algorithm
     """
-    output_img = nib.Nifti1Image(data, np.eye(4))
+    output_img = nib.nifti1.Nifti1Image(data, np.eye(4), **kwargs)
```
The `np.eye(4)` should also be an input argument, but the default should be `np.eye(data.ndim+1)` I think

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -58,7 +61,20 @@ def save_nifti_3d(data, output_file):
+        rng = np.random.RandomState(42)
+        fit = OsipiBase(algorithm=args.algorithm)
+        S0 = 1
+        gd = GenerateData(rng=rng)
```
No need to generate data, that's just for testing.

Unique-Usman replied:
Hello, I do not quite understand, how do toget the signal value. Also, what are the typical algorithm that can be passed as argument. Also, where is the test file tested https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/blob/main/tests/IVIMmodels/unit_tests/test_ivim_synthetic.py. Thank you.

etpeterson replied:
The signal comes from the loaded nifti data. The testing is entirely different and uses synthesized data. It's just an example of how to use the OsipiBase class.

The algorithms are listed [here](https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/blob/main/doc/code_contributions_record.csv), but not all are wrapped yet.

The test file is run when you run `pytest`, but specifically the slow tests as is done in the [analysis](https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/blob/main/.github/workflows/analysis.yml) workflow.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -58,7 +61,20 @@ def save_nifti_3d(data, output_file):
+        # signal = gd.ivim_signal(D, Dp, f, S0, bvals, SNR, rician_noise)
+
         # Passing the values to the selectect algorithm and saving it
+        [f_fit, Dp_fit, D_fit] = fit.osipi_fit(signal, bvals)
```
@IvanARashid had a preliminary wrapper that would make this work, but perhaps until that's ready we can loop over the point. Like this.

```
def loop_over_first_n_minus_1_dimensions(arr):
  """
  Loops over the first n-1 dimensions of a numpy array.

  Args:
      arr: A numpy array.

  Yields:
      A tuple containing the indices for the current iteration and a flattened view of the remaining dimensions.
  """
  n = arr.ndim
  for idx in np.ndindex(*arr.shape[:n-1]):
    flat_view = arr[idx].flatten()
    yield idx, flat_view


f_image = np.zeros_like(data.shape[:data.ndim-1]
D_image = np.zeros_like(data.shape[:data.ndim-1]
Dp_image = np.zeros_like(data.shape[:data.ndim-1]
for idx, view in loop_over_first_n_minus_1_dimensions(data):
    [f_fit, Dp_fit, D_fit] = fit.osipi_fit(view, bvals)
    f_image[idx]=f_fit
    Dp_image[idx]=Dp_fit
    D_image[idx]=D_fit
```

Unique-Usman replied:
Hello, how does this come in ?

etpeterson replied:
This shows how to loop over the data. `data` being loaded from the 4D image. The last dimension is the signal to fit.

Unique-Usman replied:
Also, how do I pass the value for the signal in `[f_fit, Dp_fit, D_fit] = fit.osipi_fit(signal, bvals)`

Unique-Usman replied:
Got it, thank you.

Unique-Usman replied:
> @IvanARashid had a preliminary wrapper that would make this work, but perhaps until that's ready we can loop over the point. Like this.
> 
> ```
> def loop_over_first_n_minus_1_dimensions(arr):
>   """
>   Loops over the first n-1 dimensions of a numpy array.
> 
>   Args:
>       arr: A numpy array.
> 
>   Yields:
>       A tuple containing the indices for the current iteration and a flattened view of the remaining dimensions.
>   """
>   n = arr.ndim
>   for idx in np.ndindex(*arr.shape[:n-1]):
>     flat_view = arr[idx].flatten()
>     yield idx, flat_view
> 
> 
> f_image = np.zeros_like(data.shape[:data.ndim-1]
> D_image = np.zeros_like(data.shape[:data.ndim-1]
> Dp_image = np.zeros_like(data.shape[:data.ndim-1]
> for idx, view in loop_over_first_n_minus_1_dimensions(data):
>     [f_fit, Dp_fit, D_fit] = fit.osipi_fit(view, bvals)
>     f_image[idx]=f_fit
>     Dp_image[idx]=Dp_fit
>     D_image[idx]=D_fit
> ```
I was able to make it run by using the np.getfromtxt as you made mention and also using another approach from this but similar.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,80 @@
+
+    return json_data
+
+def save_nifti_3d(data, output_file, **kwargs):
```
Let's not necessarily limit it to 3D

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,80 @@
+        bvals = read_json_file(bval_file)
+
+        # Pass additional arguments to the algorithm
+        rng = np.random.RandomState(42)
```
Random is not needed here

etpeterson reviewed (changes requested):
You can test it with some nifti files. There are nifti files in our zenodo data repository you can download. You might need to merge them to create a 4D to fully test though.

IvanARashid commented:
Regarding the fitting of multiple voxels. Yes, I was working on a solution for this in the wrapper, although that work was stranded due to the previous git issues. I won't be able to revisit this for a few weeks, but the solution that @etpeterson suggested is a fine placeholder for now. I.e. loop over all the voxels and fit them individually. Once I've implemented this functionality into the wrapper, we can get rid of it in this function.

When this functionality is in the wrapper, the returned result will be parametric map volumes, i.e 2D or 3D images of parameter values. So you could try to implement something similar now. Or we could hold this PR until multi voxel fitting is in the wrapper.

etpeterson commented:
> You can test it with some nifti files. There are nifti files in our zenodo data repository you can download. You might need to merge them to create a 4D to fully test though.

I would definitely recommend testing it like this, I think it would make things much more clear. It looks like the abdomen or the brain nifti files (`nii.gz`) look like they're already correct with an associated bval file.
https://zenodo.org/records/10696605

Unique-Usman commented:
> > You can test it with some nifti files. There are nifti files in our zenodo data repository you can download. You might need to merge them to create a 4D to fully test though.
> 
> I would definitely recommend testing it like this, I think it would make things much more clear. It looks like the abdomen or the brain nifti files (`nii.gz`) look like they're already correct with an associated bval file. https://zenodo.org/records/10696605

Thank for the review, the signal variable is still missing ? https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/pull/60/commits/6db42c1dcc41221c293c9f7dd347c5da6fee7f06#diff-aea94673dea889894c34fe41c94d6cb8fb0d3373e0677947893ba1e79f4359edR67

etpeterson commented:
> > > You can test it with some nifti files. There are nifti files in our zenodo data repository you can download. You might need to merge them to create a 4D to fully test though.
> > 
> > 
> > I would definitely recommend testing it like this, I think it would make things much more clear. It looks like the abdomen or the brain nifti files (`nii.gz`) look like they're already correct with an associated bval file. https://zenodo.org/records/10696605
> 
> Thank for the review, the signal variable is still missing ?

In the code I sent the `signal` variable is `view`. The 4D data is x, y, z, signal. So you want the signal dimension input into the fitting.

Unique-Usman commented:
> > > > You can test it with some nifti files. There are nifti files in our zenodo data repository you can download. You might need to merge them to create a 4D to fully test though.
> > > 
> > > 
> > > I would definitely recommend testing it like this, I think it would make things much more clear. It looks like the abdomen or the brain nifti files (`nii.gz`) look like they're already correct with an associated bval file. https://zenodo.org/records/10696605
> > 
> > 
> > Thank for the review, the signal variable is still missing ?
> 
> In the code I sent the `signal` variable is `view`. The 4D data is x, y, z, signal. So you want the signal dimension input into the fitting.

I now, understand,  I will also test the script now also, I also got to know that the algorithm are located at `src/standardized`.

etpeterson commented:
Just checking on this. Is it waiting on me?

etpeterson commented:
> > Just checking on this. Is it waiting on me?
> 
> No @etpeterson, I wanted to test it out and see it workings. I will do that. Recently had lot of different school activites.

Ok, no problem. Just wanted to make sure.

Unique-Usman commented:
@etpeterson I downloaded the Data from the link, I was trying to test it out. Checking the content of brain.bval and brain.bvec, I added the two function read_bval_file and read_bval_file as they were not json. 

`python3 -m WrapImage.nifti_wrapper brain.nii.gz brain.bvec brain.bval` 

I am current using the brain.nii.gz, but I Keep getting this error below. 

[
![Screenshot from 2024-04-10 15-14-10](https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/assets/86585626/e433be00-b461-4c2d-907f-ab171a1b5198)

Unique-Usman commented:
> @etpeterson I downloaded the Data from the link, I was trying to test it out. Checking the content of brain.bval and brain.bvec, I added the two function read_bval_file and read_bval_file as they were not json.
> 
> `python3 -m WrapImage.nifti_wrapper brain.nii.gz brain.bvec brain.bval`
> 
> I am current using the brain.nii.gz, but I Keep getting this error below.
> 
> [ ![Screenshot from 2024-04-10 15-14-10](https://private-user-images.githubusercontent.com/86585626/321374585-e433be00-b461-4c2d-907f-ab171a1b5198.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyMDc1NDUsIm5iZiI6MTcxMzIwNzI0NSwicGF0aCI6Ii84NjU4NTYyNi8zMjEzNzQ1ODUtZTQzM2JlMDAtYjQ2MS00YzJkLTkwN2YtYWIxNzFhMWI1MTk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE1VDE4NTQwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkOGRkZjYzOGFkYmQwMWRkNDQ2ZWQwMDNmZDY4MWZkMmVmZTY3NWZhOGM0NDQ2MGVjYjA2ZjAzYmM2M2MyZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1vqk32kbNXGNTAonfw-KC5QiWtBuxE5XEtlwcdR7zko)
> 
> I am using the algorithms in the src/standardized. I also changed the algorithm couple of time also, but it keep giving almost similar error.

@bringing your attention to this.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,125 @@
+        Dp_image = np.zeros_like(data.shape[:data.ndim-1])
+        for idx, view in loop_over_first_n_minus_1_dimensions(data):
+            [f_fit, Dp_fit, D_fit] = fit.osipi_fit(view, bvals)
+            f_image[idx]=f_fit
```
Spaces around the `=` in these lines for readability please.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,125 @@
+
+    with open(bval_file, "r") as f:
+        try:
+            bval_data = [int(val) for val in f.read().split()]
```
Not `int`, at least a `float`.

etpeterson commented:
> > @etpeterson I downloaded the Data from the link, I was trying to test it out. Checking the content of brain.bval and brain.bvec, I added the two function read_bval_file and read_bval_file as they were not json.
> > `python3 -m WrapImage.nifti_wrapper brain.nii.gz brain.bvec brain.bval`
> > I am current using the brain.nii.gz, but I Keep getting this error below.
> > [ ![Screenshot from 2024-04-10 15-14-10](https://private-user-images.githubusercontent.com/86585626/321374585-e433be00-b461-4c2d-907f-ab171a1b5198.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyMDc1NDUsIm5iZiI6MTcxMzIwNzI0NSwicGF0aCI6Ii84NjU4NTYyNi8zMjEzNzQ1ODUtZTQzM2JlMDAtYjQ2MS00YzJkLTkwN2YtYWIxNzFhMWI1MTk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDE1VDE4NTQwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkOGRkZjYzOGFkYmQwMWRkNDQ2ZWQwMDNmZDY4MWZkMmVmZTY3NWZhOGM0NDQ2MGVjYjA2ZjAzYmM2M2MyZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1vqk32kbNXGNTAonfw-KC5QiWtBuxE5XEtlwcdR7zko)
> > I am using the algorithms in the src/standardized. I also changed the algorithm couple of time also, but it keep giving almost similar error.
> 
> @etpeterson bringing your attention to this.

I would guess that the bvals and bvecs aren't numpy arrays or are the wrong size. Maybe try transposing them?
Perhaps also try parsing the bval and bvec files with np.genfromtxt rather than manually. https://numpy.org/doc/stable/reference/generated/numpy.genfromtxt.html

It could also be the input data are not the right size. The data length should match the bvalue length.

I'm just returning from a break so I could try to debug it later this week if you're still having trouble.

etpeterson reviewed (commented):
I think it's close

Unique-Usman commented:
@etpeterson thanks, I will look into this.

Unique-Usman commented:
![Screenshot from 2024-04-23 13-49-33](https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/assets/86585626/a7d21dd0-596f-4bef-ac51-f09a153d7398)
![Screenshot from 2024-04-23 13-48-16](https://github.com/OSIPI/TF2.4_IVIM-MRI_CodeCollection/assets/86585626/aab1f36d-8772-4ed1-abc0-69d9d970450a)

Unique-Usman commented:
@etpeterson, kindly check this out. Thanks. And also welcome back from the break.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -0,0 +1,127 @@
+        Dp_image = []
+        D_image = []
+
+        for idx, view in loop_over_first_n_minus_1_dimensions(data):
```
I tried it and didn't realize it was fitting and cancelled the run only to realize it was actually working correctly. So perhaps wrap this loop in `tqdm`? https://github.com/tqdm/tqdm

It's already a requirement so should be a small addition. If that's harder than expected, just a print before the loop saying that it's fitting with the given algorithm would be good enough too.

etpeterson reviewed (changes requested):
Looks good. Just 2 small comments. Nice work.

Unique-Usman commented:
@etpeterson I also do not why some of the test is failing, I checked it. But, it does not seem related with my changed.

etpeterson commented on WrapImage/nifti_wrapper.py:
```diff
@@ -102,7 +103,7 @@ def loop_over_first_n_minus_1_dimensions(arr):
         D_image = []
 
-        for idx, view in loop_over_first_n_minus_1_dimensions(data):
+        for idx, view in tqdm(loop_over_first_n_minus_1_dimensions(data), desc=f"{args.algorithm} is fitting", dynamic_ncols=True, total=702464):
```
Is this necessary? `total=702464` We can't hard code any numbers because we don't know the image size. If it's needed could we just extract it from the image itself?

etpeterson replied:
Was there an update to this? I don't see a change.

etpeterson reviewed (changes requested):
Just the one comment.

etpeterson commented:
> @etpeterson I also do not know why some of the test is failing, I checked it. But, it does not seem related with my changes.

Not your fault, it looks like github changed something that broke older python versions.
https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json

I'm not sure if it's going to be fixed or not, but let's leave it for a day or two to see if it comes back.
https://github.com/actions/setup-python/issues/850

etpeterson commented:
> > @etpeterson I also do not know why some of the test is failing, I checked it. But, it does not seem related with my changes.
> 
> Not your fault, it looks like github changed something that broke older python versions. https://raw.githubusercontent.com/actions/python-versions/main/versions-manifest.json
> 
> I'm not sure if it's going to be fixed or not, but let's leave it for a day or two to see if it comes back. [actions/setup-python#850](https://github.com/actions/setup-python/issues/850)

Ok, this was fixed. Just the one change that I see remaining.

Unique-Usman commented:
@etpeterson, the tqdm works by having the information about the number of iteration beforehand. But, in our code. We are using yield function which would not give the information about the number of iteration beforehand. The progress bar will not be displayed if the number of iteration is not known. The solution is to provide a number(if known). Also, if we can not get the number of iteration. We can basically print when it is starting and when it is ending.

etpeterson commented:
> @etpeterson, the tqdm works by having the information about the number of iteration beforehand. But, in our code. We are using yield function which would not give the information about the number of iteration beforehand. The progress bar will not be displayed if the number of iteration is not known. The solution is to provide a number(if known). Also, if we can not get the number of iteration. We can basically print when it is starting and when it is ending. 

We do know the number, it's almost the same code as the generator.
`len(arr.shape[:n-1]))`

Also, it looks like if you wrapped it up into a class it could query that implicitly. https://stackoverflow.com/questions/41985993/tqdm-show-progress-for-a-generator-i-know-the-length-of

Unique-Usman commented:
@etpeterson, I already make the value passed to the total argument dynamic.

etpeterson commented:
Looks good to me, thanks!
